Concepts section overhaul (mirror project) (#296)
pinot-contrib/pinot-docs: basics/concepts/time-boundary.md

---

> description: In this Apache Pinot concepts guide, we'll learn about time boundaries in hybrid tables.

Reviewer: ```suggestion
description: Learn about time boundaries in hybrid tables.
```

---

> # Time Boundary

Reviewer: ```suggestion
# Time boundary
```

> In this guide we'll learn about time boundaries in hybrid tables.
> Hybrid tables are when we have offline and real-time tables with the same name.

Reviewer: ```suggestion
Learn about time boundaries in hybrid tables.
Hybrid tables are when we have offline and real-time tables with the same name.
```

Reviewer: Hybrid tables occur when you create an offline and real-time table with the same name?

> When querying these tables, the Pinot Broker needs to decide which records to read from the offline table and which to read from the real-time table.

Reviewer: ```suggestion
When querying these tables, the Pinot broker decides which records to read from the offline table and which to read from the real-time table.
```

> When querying these tables, the Pinot Broker needs to decide which records to read from the offline table and which to read from the real-time table.
> It does this using the time boundary.

Reviewer: ```suggestion
When querying hybrid tables, the Pinot broker decides which records to read from the offline table and which to read from the real-time table by using the time boundary.
```

## How is the time boundary determined?

The time boundary is determined by looking at the maximum end time of the offline segments and the segment ingestion frequency specified for the offline table.

If it's set to hourly, then:

```
timeBoundary = Maximum end time of offline segments - 1 hour
```

Otherwise: 

```
timeBoundary = Maximum end time of offline segments - 1 day
```

It is possible to force the hybrid table to use max(all offline segments' `end time`) by calling the API (V 0.12.0+)

```
curl -X POST \
  "http://localhost:9000/tables/{tableName}/timeBoundary" \
  -H "accept: application/json"
```

> Note that this will not automatically update the timeboundary as more segments are added to the offline table and must be called each time a segment with more recent end time is uploaded to the offline table. Users can revert back to using the derived timeboundary by calling API:

Reviewer: ```suggestion
Note that this will not automatically update the time boundary as more segments are added to the offline table, and must be called each time a segment with more recent end time is uploaded to the offline table. You can revert back to using the derived time boundary by calling API:
```

```
curl -X DELETE \
  "http://localhost:9000/tables/{tableName}/timeBoundary" \
  -H "accept: application/json"
```

## Querying

> When a query for a hybrid table is received by a Pinot Broker, the broker sends a time boundary annotated version of the query to the offline and real-time tables. 

Reviewer: ```suggestion
When a Pinot broker receives a query for a hybrid table, the broker sends a time boundary annotated version of the query to the offline and real-time tables. 
```
